sys/kern/rman.c

```diff
@@ -1,83 +1,252 @@
+/*
+ * The kernel resource manager.
+ *
+ * Based on FreeBSD `subr_rman.c` file.
+ *
+ * This code is responsible for keeping track
+ * of hardware resources which are apportioned out to various drivers.
+ * It does not actually assign those resources, and it is not expected
+ * that end-device drivers will call into this code directly.  Rather,
+ * the code which implements the buses that those devices are attached to,
+ * and the code which manages CPU resources, will call this code, and the
+ * end-device drivers will make upcalls to that code to actually perform
+ * the allocation.
+ */
+
 #include <sys/mimiker.h>
 #include <sys/rman.h>
 #include <sys/malloc.h>
 
-static KMALLOC_DEFINE(M_RES, "resources");
+/*
+ * Bits 15-10 of the `flags` argument to `rman_reserve_resource`
+ * ensure alignment to 2^bits[15-10].
+ */
 
-static bool rman_find_gap(rman_t *rm, rman_addr_t *start_p, rman_addr_t end,
-                          size_t count, size_t bound, resource_t **res_p) {
-  assert(mtx_owned(&rm->rm_lock));
+/* Alignment bits start at bit 10. */
+#define RMAN_ALIGNMENT_SHIFT 10
 
-  rman_addr_t start = *start_p;
+/* Positions the alignment bits for the `flags` argument. */
+#define RMAN_ALIGNMENT_ENCODE(x) ((x) << RMAN_ALIGNMENT_SHIFT)
 
-  if (end < rm->rm_start || start > rm->rm_end)
-    return false;
+/* Alignment mask to extract the alignment bits. */
+#define RMAN_ALIGNMENT_MASK (0x003F << RMAN_ALIGNMENT_SHIFT)
+
+/* Extracts alignment bits. */
+#define RMAN_ALIGNMENT_EXTRACT(x)                                              \
+  (((x)&RMAN_ALIGNMENT_MASK) >> RMAN_ALIGNMENT_SHIFT)
+
+/* Decodes alignment bits from the `flags` argument. */
+#define RMAN_ALIGNMENT_GET(x) ((1ull << RMAN_ALIGNMENT_EXTRACT(x)) - 1)
+
+#define RMAN_ALIGN(addr, amask) (((addr) + (amask)) & ~(amask))
+
+#define RESOURCE_GET_RMAN(r) ((r)->r_rg->rg_rman)
+
+typedef TAILQ_HEAD(, resource) res_list_t;
+
+struct rman_region {
+  rman_t *rg_rman;                  /* resource manager */
+  rman_addr_t rg_start;             /* first physical address */
+  rman_addr_t rg_end;               /* last physical address */
+  res_list_t rg_resources;          /* all managed resources */
+  TAILQ_ENTRY(rman_region) rg_link; /* link on resource manager list */
+};
+
+static KMALLOC_DEFINE(M_RES, "resources & regions");
+
+void rman_init(rman_t *rm, const char *name) {
+  rm->rm_name = name;
+  mtx_init(&rm->rm_lock, 0);
+  TAILQ_INIT(&rm->rm_regions);
+}
+
+void rman_manage_region(rman_t *rm, rman_addr_t start, rman_addr_t end) {
+  rman_region_t *rg, *s, *t;
+
+  rg = kmalloc(M_RES, sizeof(rman_region_t), M_ZERO);
```

Comment:
Please notice that we to the allocation here because we want to avoid calling `kmalloc` (without `M_NOWAIT`) inside `WITH_MTX_LOCK (&rm->rm_lock) { ...}`.

Reply:
Ok, fair enough. Allocating memory for resource in before, just in case it's going to be used seems to be a better choice than calling kmalloc and potentially going to sleep while holding a mutex.